### PR TITLE
shared library support (issue #6)

### PR DESCRIPTION
**Pull request by stefantalpalaru**
and build options for shared/static libs (both enabled by default)

### CODE REVIEW

**Comment by jbkempf** [comment]
Looks go to me.

@chouquette an opinion on the matter?

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,9 +1,16 @@
 cmake_minimum_required(VERSION 3.1)
-project(spatialaudio VERSION 0.3.0)
+
+set(PACKAGE_VERSION_MAJOR "0")
+set(PACKAGE_VERSION_MINOR "3")
+set(PACKAGE_VERSION_PATCH "0")
+project(spatialaudio VERSION "${PACKAGE_VERSION_MAJOR}.${PACKAGE_VERSION_MINOR}.${PACKAGE_VERSION_PATCH}")
```
Comment (chouquette) [inline CMakeLists.txt]: I would argue this needs to be in a different commit

File: CMakeLists.txt [hunk 2]
```diff
@@ -86,12 +108,11 @@ if(${HAVE_MYSOFA})
 endif(${HAVE_MYSOFA})
 
 configure_file(
-  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/spatialaudio.pc.cmake"
-  "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.pc"
-  @ONLY
+    "${CMAKE_CURRENT_SOURCE_DIR}/cmake/spatialaudio.pc.cmake"
+    "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.pc"
+    @ONLY
```
Comment (chouquette) [inline CMakeLists.txt]: This should also be in a separated commit

File: CMakeLists.txt [hunk 3]
```diff
@@ -67,11 +74,26 @@ if(${MYSOFA_FOUND})
     include_directories(${MYSOFA_INCLUDE_DIRS})
 endif(${MYSOFA_FOUND})
 
-add_library(spatialaudio STATIC ${sources})
+if(BUILD_STATIC_LIBS)
+    add_library(spatialaudio-static STATIC ${sources})
+    if(${MYSOFA_FOUND})
+        target_link_libraries(spatialaudio-static ${MYSOFA_LIBRARIES})
```
Comment (chouquette) [inline CMakeLists.txt]: Is this required? I'd expect sources only to affect the resulting static library (granted, my cmake skillz are... let's say rusted)

File: CMakeLists.txt [hunk 4]
```diff
@@ -67,11 +74,26 @@ if(${MYSOFA_FOUND})
...
-endif(${MYSOFA_FOUND})
+if(BUILD_SHARED_LIBS)
+    add_library(spatialaudio-shared SHARED ${sources})
+    if(${MYSOFA_FOUND})
+        target_link_libraries(spatialaudio-shared ${MYSOFA_LIBRARIES})
+    endif(${MYSOFA_FOUND})
+    SET_TARGET_PROPERTIES(spatialaudio-shared PROPERTIES OUTPUT_NAME spatialaudio CLEAN_DIRECT_OUTPUT 1)
+    set_property(TARGET spatialaudio-shared PROPERTY VERSION "${PACKAGE_VERSION_MAJOR}.${PACKAGE_VERSION_MINOR}.${PACKAGE_VERSION_PATCH}")
+    set_property(TARGET spatialaudio-shared PROPERTY SOVERSION ${PACKAGE_VERSION_MAJOR} )
+    set_property(TARGET spatialaudio-shared PROPERTY C_VISIBILITY_PRESET hidden)
```
Comment (chouquette) [inline CMakeLists.txt]: Is this property also used for C++?

File: CMakeLists.txt (hunk 2 above)
Comment (chouquette) [reply CMakeLists.txt]: Also wouldn't it make sense to ship 2 .pc files now? (To be fair I'm not sure what the recommended/idiomatic way to do this is)

**Comment by jbkempf** [comment]
@stefantalpalaru did you see @chouquette comments?

**Comment by stefantalpalaru** [comment]
> @stefantalpalaru did you see @chouquette comments?

Yes, of course, but I'm not playing the bike-shedding game. Mine is a serious contribution. If you want to argue about stuff like number of commits or want somebody to teach you CMake/pkg-config/dynamic libraries/linking/etc. because you can't be bothered to do your own research, please waste somebody else's time.

If you have serious questions, I'll be happy to try and answer them.

**Comment by jbkempf** [comment]
2 remarks suggesting a split which are trivial to do (takes a couple of minutes), and one legitimate question (it's clearly not obvious that C_VISIBILITY_PRESET applies to C++ also, even after reading the documentation) are quite far from bike-shedding. Especially if it is only the first question on the discussion.

**Comment by stefantalpalaru** [comment]
Nobody benefits from rewriting Git history to split commits. It is an absurd request for something that helps no one and damages me as a contributor by asking me to waste my time and mess up the proper record of my contribution. It would be like me asking you to delete your previous comment in this thread and split it in two new ones.

> it's clearly not obvious that C_VISIBILITY_PRESET applies to C++ also

I don't think it does. Try also setting the CXX_VISIBILITY_PRESET property and run make with VERBOSE=1 to see the resulting "-fvisibility=hidden" argument to the compiler.

**Comment by jbkempf** [comment]
> Nobody benefits from rewriting Git history to split commits. It is an absurd request for something that helps no one

Absolutely not. One feature per commit helps reviewing, reading history, reverting commits and bisecting. A lot of projects require that, and a lot of PR are using multiple commits.

> and damages me as a contributor by asking me to waste my time and mess up the proper record of my contribution.

Of course not! Instead of 1 commit in the tree, you got 2 consecutive ones. It even increases your commit count, if you care about that. I don't see how that "messes up the record". And I clearly don't see why this requests upsets you so much.

> It would be like me asking you to delete your previous comment in this thread and split it in two new ones.

That is a ridiculous analogy. The idea for multiple commits if for other people, in the future, to read the code and understand it. The comments here are a process that is not necessary to understand the code.

Anyway, back to the point. This is clearly not bike-shedding, like asking you to remove trailing spaces or aligning equal signs vertically...

> me to waste my time

Finally, in the time of your comments, you could have done it already.

But anyway...

> I don't think it does. Try also setting the CXX_VISIBILITY_PRESET property and run make with VERBOSE=1 to see the resulting "-fvisibility=hidden" argument to the compiler.

Which is exactly why @chouquette asked you: "Why do you set a C property on a C++ project?".

It is a very good question, is not obvious at all, and is clearly not bike-shedding.

Yes, I think you are clearly over-reacting on this.

**Comment by stefantalpalaru** [comment]
I tested it myself and it works as I guessed it would. New property added to this pull request.

**Comment by stefantalpalaru** [comment]
>  It even increases your commit count, if you care about that.

I don't. I care about my time being wasted, since it is a finite resource.

> I don't see how that "messes up the record".

You don't see how deleting a commit from 25 days ago and creating 3 new ones now changed the history of when my initial contribution was created?

> The idea for multiple commits if for other people, in the future, to read the code and understand it.

I agree with the concept when a single commit is too big to be read in 2-3 minutes. My commit is very small and, surely, easy to grok.

> "Why do you set a C property on a C++ project?"

Your C++ project contains 3 C files.

> It is a very good question, is not obvious at all, and is clearly not bike-shedding.

OK, point taken, question addressed, pull request improved.

**Comment by stefantalpalaru** [comment]
CXX_VISIBILITY_PRESET change reverted, because it won't work without marking exported symbols with ```__attribute__ ((visibility ("default")))``` and this is too big of a change for this pull request. Keep it in mind for when you have a significant amount of C++ symbols that can be safely hidden.

As an alternative, you can mark just the hidden ones and not change the compiler's default visibility policy.

Hiding all the C symbols does not create problems for VLC, so I'm assuming those are not part of the public API.

**Comment by chouquette** [comment]
Merged. Thanks for your patch
